### PR TITLE
Add unit tests for render engines

### PR DESCRIPTION
**Pull request by tomers**
Adds initial framework to run pytest tests.
Test text render engine.

### CODE REVIEW

**Comment by maresb** [comment]
Woah, this is really awesome!!!

I know from the draft state that you're not done. My main thought is that we shouldn't write files into the repo; instead we should use the `tmp_path` fixture.

My other thought is that I wonder if there's a simple way to automate regenerating the images (e.g. if we change the default font), but that's less important.

**Comment by maresb** [comment]
Ah, and because of Python import weirdness, you should add `import PIL.ImageOps` in addition to `import PIL` at the top of `test_render_engines.py`.

**Comment by maresb** [comment]
@tomers, you should probably rebase this on the latest main

**Comment by maresb** [comment]
I don't like running pytest in pre-commit. Individual commits should be able to break tests. For example, this is desirable in test-driven development where you first commit a broken test and then fix it in the subsequent commits.

**Comment by maresb** [comment]
I rely on pre-commit for linting and type-checking which I find essential for lowering my commit entropy.

Including pytest will encourage developers to make larger commits which makes review difficult.

Also, I can't think of a single major open-source project that runs pytest in pre-commit. I think there are good reasons for that.

**Comment by tomers** [comment]
> My main thought is that we shouldn't write files into the repo; instead we should use the tmp_path fixture.
I prefer not to. The reason for this is that:
1. I delete each created file when the test succeed
2. In case of test failure, the file is in the directory just above the reference `expected_renders` directory, thus it is easy to compare. In case the new file is the correct one (e.g. following a code change), we can just move the file into that directory. It is very easy to do from the IDE, where the actual image file is near the code, the test code, and the reference image file, and by simple and short drag-and-drop action it can be moved to the `expected_renders` directory, and be committed.

> My other thought is that I wonder if there's a simple way to automate regenerating the images (e.g. if we change the default font), but that's less important.
I don't think we should. I prefer to do it manually, and examine each change. As I wrote above, we can just drag and drop in order to move the new version of the file to its designated place.

> I rely on pre-commit for linting and type-checking which I find essential for lowering my commit entropy.
> Including pytest will encourage developers to make larger commits which makes review difficult.
> Also, I can't think of a single major open-source project that runs pytest in pre-commit. I think there are good reasons for that.
Ok, I agree to your reasoning. I've removed it. Instead, I've added it to the tox cofiguration in `pyproject.toml`. Please review this, and see if I did it in a proper manner. I want to allow developers to type a short command (e.g. `tox` to run all tests, including generation of coverage reports and any other artifacts of a test run).

**Comment by maresb** [comment]
@tomers, I just started to review. Are you still making changes?

**Comment by maresb** [comment]
Marking as draft for now, just click "Ready for review" when you'd like me to take a look.

**Comment by tomers** [comment]
This PR is becoming pretty big, so I'll stop here, and submit my future changes in a separate PR

File: src/labelle/lib/render_engines/tests/test_render_engines.py
Comment (maresb) [inline src/labelle/lib/render_engines/tests/test_render_engines.py:97]: Hmm, this reminds me: for the last patch release I added a hack where the empty string for code128 gets replaced by a space character to avoid an exception. Is this how we want to handle it? We should probably add another test case here for `(BarcodeType.CODE128, " ")`.

**Review by maresb — CHANGES_REQUESTED** [review]
I haven't looked in detail about why `= None` is necessary for the rendering init, but it stands out to me as odd. Probably we shouldn't need to weaken function signatures in order to make tests easier. Maybe add a kwargs to the parameterize?

File: src/labelle/lib/render_engines/barcode.py
Comment (tomers) [reply src/labelle/lib/render_engines/barcode.py]: What is the purpose of default barcode type then?
This is API for developers, and in the future one might costruct his own code which directly uses these classes. I don't see a point in requiring anyone to explicitly provide values for arguments we have default for in the implemnentation.

File: src/labelle/lib/render_engines/barcode_with_text.py
Comment (tomers) [reply src/labelle/lib/render_engines/barcode_with_text.py:22]: Same here

File: src/labelle/lib/render_engines/barcode.py
Comment (maresb) [reply src/labelle/lib/render_engines/barcode.py]: In that case shouldn't it be `barcode_type: BarcodeType = DEFAULT_BARCODE_TYPE`?

**Comment by tomers** [comment]
> It's a lot more conventional to put `tests/` alongside `src/`. I don't feel particularly strongly about this, but unless there's a particular reason it may be better in terms of standardization to move it there.

I think it would be better to have separate `tests` directory in each relevant code section, e.g. render engine tests under `src/labelle/lib/render_engines/tests`. This will prevent building of similar hierarchy under a global `tests` directory, which might break in the future, if things move along.

File: src/labelle/lib/render_engines/exceptions.py
Comment (tomers) [reply src/labelle/lib/render_engines/exceptions.py:4]: I think we are covered:
```
❯ labelle --output=console
Usage: labelle [OPTIONS] [TEXT]... COMMAND [ARGS]...
Try 'labelle --help' for help.
╭─ Error ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ Invalid value: No elements to print                                                                                                                                                   │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

❯ labelle --output=console
Usage: labelle [OPTIONS] [TEXT]... COMMAND [ARGS]...
Try 'labelle --help' for help.
╭─ Error ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ Invalid value: No elements to print                                                                                                                                                   │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

❯ labelle --output=console --picture  
╭─ Error ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ Option '--picture' requires an argument.                                                                                                                                              │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

❯ labelle --output=console --qr
╭─ Error ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ Option '--qr' requires an argument.                                                                                                                                                   │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

❯ labelle --output=console --barcode
╭─ Error ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ Option '--barcode' requires an argument.                                                                                                                                              │
╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

File: src/labelle/lib/render_engines/tests/test_render_engines.py
Comment (tomers) [reply src/labelle/lib/render_engines/tests/test_render_engines.py:38]: oh! that's interesting. It should be omitted, and the default is then `threshold=0.001`.
The thing is, that in CI, for some unknown reason, I get different bitmaps than on my development machine. There's small fluctuations there.
The `PIL` library is the same in my host and in CI.
This needs to be investigated further. I will be grateful if you could take a look at that. For the meanwhile, I would keep it as is, temporary.

File: src/labelle/lib/render_engines/text.py
Comment (tomers) [reply src/labelle/lib/render_engines/text.py:18]: See above

File: src/labelle/lib/render_engines/barcode.py
Comment (tomers) [reply src/labelle/lib/render_engines/barcode.py]: It is a matter of convention. We got this:
```
self.barcode_type = barcode_type or DEFAULT_BARCODE_TYPE
```
I think it is easier on the eyes to see `None` as default elsewhere, and then have some logic to align things.
wdyt?

Comment (maresb) [reply src/labelle/lib/render_engines/barcode.py]: The `= None` default pattern is usually only used for mutable types like lists or dicts. In the mutable case, a single mutable instance is be created on module load and then reused for all invocations. :skull: 

It should be pretty rare that you actually need a mutable default, so the common convention is to set the default value to be the default value unless mutable is necessary.

File: src/labelle/lib/render_engines/exceptions.py
Comment (maresb) [reply src/labelle/lib/render_engines/exceptions.py:4]: But just `labelle` without `--output` prints a blank label for me.

File: src/labelle/cli/cli.py
Comment (maresb) [reply src/labelle/cli/cli.py:162]: Hmm, maybe I'm confused, but I think you did the wrong thing in 0741930.

File: src/labelle/lib/render_engines/tests/test_render_engines.py
Comment (maresb) [reply src/labelle/lib/render_engines/tests/test_render_engines.py:40]: Strange. Created #39

File: src/labelle/cli/cli.py
Comment (tomers) [reply src/labelle/cli/cli.py:162]: Indeed! good catch. I've fixed it.

**Comment by tomers** [comment]
> I haven't looked in detail about why = None is necessary for the rendering init, but it stands out to me as odd.
> Probably we shouldn't need to weaken function signatures in order to make tests easier. Maybe add a kwargs to the parameterize?

In `cli.py` we *explicitly* pass these arguments, which might be `None`, thus we must support the case where the argument value is `None`, and handle it in the function body, instead of in the declaration.

**Comment by maresb** [comment]
This still seems fishy to me. Why is the value `None` in the first place? If they're coming from `cli.py` then perhaps our defaults there are incorrect. For example, I think this is wrong:

https://github.com/labelle-org/labelle/blob/d75a71e63fb2864d2105b93f926c1ef4bb9b46be/src/labelle/cli/cli.py#L214

Shouldn't it be this?
```python
] = DEFAULT_BARCODE_TYPE
```

Also, I'm not sure if you saw this remark, or if it's hidden for you: https://github.com/labelle-org/labelle/pull/32#discussion_r1584773683

Thanks for all your work on this!

**Comment by tomers** [comment]
@maresb I've fixed the default handling of barcode type.
If possible, please review this and merge.
I got lost amongs the different PRs that we got. I'll leave it to you to merge.
I guess I'll have some more breaking changes coming up soon. Would it be ok that you release a new major version with breaking changes, and then we introduce more breaking changes, or go you prefer that we keep work on `main` branch, without releasing, and then release once we're done?

File: src/labelle/lib/render_engines/text.py
Comment (maresb) [reply src/labelle/lib/render_engines/text.py:18]: Here is another instance of a strange unnecessary default argument of `None`.

File: src/labelle/lib/render_engines/barcode_with_text.py
Comment (maresb) [reply src/labelle/lib/render_engines/barcode_with_text.py:22]: Also here

**Comment by maresb** [comment]
@tomers, would you be able to address the remaining two default arguments? It's pretty minor, so please just merge yourself afterwards.

My wish would be to merge this PR, then #37, then #25, make one or more patch releases, and then start with the breaking changes. Would that work well with your plans?

**Comment by tomers** [comment]
@maresb, I am blocked from merging this PR. I guess it is because you're the reviewer. Please merge.

**Comment by maresb** [comment]
> I am blocked from merging this PR.

That's annoying, sorry about that! I don't want that to happen.

Looking under https://github.com/labelle-org/labelle/settings/branch_protection_rules/49111150, I see
image

I think that's what's preventing you from merging as a maintainer. I'd prefer to keep this normally checked in order to avoid accidents, but please feel free to temporarily remove this protection in cases like this.

**Comment by maresb** [comment]
Unchecking that box gives maintainers the following override:
image

**Review by maresb — APPROVED** [review]

**Pull request merged**
